### PR TITLE
Migrate TOSA input conversion to a compiler plugin.

### PR DESCRIPTION
**Pull request by ScottTodd**
Progress on https://github.com/openxla/iree/issues/15468

Using the plugin API lets us avoid plumbing `#ifdef IREE_HAVE_TOSA_INPUT` across the source tree. The `IREE_INPUT_TOSA` CMake define now just chooses whether to include the plugin or not.

Future PRs will migrate other components like the StableHLO input conversion. I chose to start with the TOSA conversion because it has fewer files.

## Build system details

This is _mostly_ a code move. The Bazel and CMake configurations were a little tricky though. Note that some CMake files here were authored manually, not via bazel_to_cmake.

The Torch input conversion has files located at `compiler/plugins/input/Torch/torch-iree/**`, with includes rooted on `torch-iree` (e.g. `#include "torch-iree/InputConversion/Passes.h"`). However, Bazel wants include paths to match the (unambiguous) `WORKSPACE`-relative paths (e.g. `#include "compiler/plugins/input/Torch/torch-iree/InputConversion/Passes.h"`). Solutions for working around that in Bazel are [documented here](https://bazel.build/tutorials/cpp-use-cases#add-include-paths).

Similarly, the Torch input conversion roots its CMake targets at `torch-iree::InputConversion`. In cases where we use relative targets that isn't an issue, but bazel_to_cmake does not understand that for absolute target labels.

We have a few options:

1. Use WORKSPACE-relative paths and generate CMake files from Bazel files
2. (this PR) Use plugin-relative paths and manually author a few CMake files
3. Use plugin-relative paths and teach bazel_to_cmake about include copts
4. Use something like `//compiler/src:defs` to propagate include paths (I tried this a few ways already)
5. Use relative include paths (`./...`)

## Future work

* This new plugin depends on `iree::compiler::Dialect::Flow::Transforms` for `IREE::Flow::createStripSignednessPass()`. I think this is the only use of that pass, so it could be moved in to the plugin.
* Mixed input dialect conversion is worth trying now

### CODE REVIEW

File: compiler/plugins/input/TOSA/tosa-iree/InputConversion/CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,9 @@ iree_tablegen_library(
 iree_cc_library(
   NAME
     PassHeaders
+  COPTS
+    "-Icompiler/plugins/input/TOSA"
+    "-I$(GENDIR)/compiler/plugins/input/TOSA"
```
Comment (ScottTodd) [inline compiler/plugins/input/TOSA/tosa-iree/InputConversion/CMakeLists.txt:27]: These are no-ops (apparently/hopefully) in CMake. Could drop them in the bazel_to_cmake conversion

Includes are already handled in CMake with
```cmake
add_library(tosa-iree_compiler_defs INTERFACE)
target_include_directories(tosa-iree_compiler_defs
  INTERFACE
  ${CMAKE_CURRENT_SOURCE_DIR}
  ${CMAKE_CURRENT_BINARY_DIR}
)

# Configures all iree_cc_* targets to take this implicit dep,
# which provides common includes and copts for the tree.
set(IREE_IMPLICIT_DEFS_CC_DEPS tosa-iree_compiler_defs)
```

this is sort of the Bazel version of that (could maybe build _that_ into the converter, with a helper CMake function mapping to some generated Bazel... idk)

File: compiler/plugins/input/TOSA/tosa-iree/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+iree_compiler_cc_library(
+    name = "registration",
+    srcs = [
+        "PluginRegistration.cpp",
+    ],
+    copts = ["-Icompiler/plugins/input/TOSA"],
+    deps = [
+        "//compiler/plugins/input/TOSA/tosa-iree/InputConversion",
```
Comment (ScottTodd) [inline compiler/plugins/input/TOSA/tosa-iree/BUILD.bazel:27]: This target corresponds to `tosa-iree::InputConversion::InputConversion` in CMake, which bazel_to_cmake can't currently handle

**Review by stellaraccident — APPROVED** [review]
Nice. Could you double check that the lit tests that moved are actually running?

**Comment by ScottTodd** [comment]
> Nice. Could you double check that the lit tests that moved are actually running?

Bazel: https://github.com/openxla/iree/actions/runs/6817676816/job/18541690731?pr=15495#step:4:677
```
//compiler/plugins/input/TOSA/tosa-iree/InputConversion/test:convert_i48_to_i64.mlir.test PASSED in 0.6s
//compiler/plugins/input/TOSA/tosa-iree/InputConversion/test:tosa_to_linalg_ext.mlir.test PASSED in 0.6s
//compiler/plugins/input/TOSA/tosa-iree/InputConversion/test:verify_compiler_tosa_input_legality.mlir.test PASSED in 0.8s
```

CMake: https://github.com/openxla/iree/actions/runs/6817676816/job/18541913066?pr=15495#step:6:1185
```
          Start  792: tosa-iree/InputConversion/test/convert_i48_to_i64.mlir.test
 517/1417 Test  #718: iree/tests/e2e/tosa_ops/check_vulkan-spirv_vulkan_logical_right_shift.mlir .................................   Passed    0.27 sec
          Start  793: tosa-iree/InputConversion/test/tosa_to_linalg_ext.mlir.test
 518/1417 Test  #724: iree/tests/e2e/tosa_ops/check_vulkan-spirv_vulkan_mul_shift.mlir ...........................................   Passed    0.25 sec
          Start  794: tosa-iree/InputConversion/test/verify_compiler_tosa_input_legality.mlir.test
```

VSCode CMake Test Explorer:
image

**Comment by ScottTodd** [comment]
Ah, I should also update CODEOWNERS for the new paths: https://github.com/openxla/iree/blob/3ef0ea1127c196029dfbd8764c94412778d592b7/.github/CODEOWNERS#L75-L77

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -73,8 +73,8 @@
 /compiler/src/iree/compiler/Dialect/Vulkan/ @antiagainst
 /compiler/src/iree/compiler/GlobalOptimization/ @hanhanW
 /compiler/src/iree/compiler/InputConversion/ @MaheshRavishankar @stellaraccident
-/compiler/src/iree/compiler/InputConversion/MHLO @hanhanW @MaheshRavishankar @rsuderman
-/compiler/src/iree/compiler/InputConversion/TOSA @MaheshRavishankar @rsuderman
+/compiler/src/iree/compiler/InputConversion/StableHLO/ @hanhanW @MaheshRavishankar @rsuderman
+/compiler/plugins/input/TOSA/ @MaheshRavishankar @rsuderman
```
Comment (ScottTodd) [inline .github/CODEOWNERS:77]: @MaheshRavishankar / @rsuderman do either of you want to weigh in on this PR? You're listed in CODEOWNERS
(I plan to migrate StableHLO next)

**Pull request merged**